File: python_sdk/services/service_discovery.py
"""
Service discovery client for PyWatt modules.

This module provides functionality for registering service providers and
discovering services through the orchestrator.
"""

import asyncio
from dataclasses import dataclass
from enum import Enum
from typing import Dict, List, Optional, Any

from ..communication import TcpChannel, ConnectionConfig, MessageError
from ..communication.message import Message, EncodedMessage


class ServiceType(Enum):
    """Types of services that can be registered."""
    HTTP = "http"
    GRPC = "grpc"
    WEBSOCKET = "websocket"
    CUSTOM = "custom"


@dataclass
class ServiceProviderInfo:
    """Information about a service provider."""
    id: str
    name: str
    service_type: ServiceType
    endpoint: str
    version: str
    metadata: Optional[Dict[str, str]] = None
    health_check_endpoint: Optional[str] = None
    
    @classmethod
    def new(cls, id: str, name: str, service_type: ServiceType, endpoint: str, version: str) -> 'ServiceProviderInfo':
        """Create a new service provider info."""
        return cls(
            id=id,
            name=name,
            service_type=service_type,
            endpoint=endpoint,
            version=version
        )
    
    def with_metadata(self, key: str, value: str) -> 'ServiceProviderInfo':
        """Add metadata to the service provider."""
        if self.metadata is None:
            self.metadata = {}
        self.metadata[key] = value
        return self
    
    def with_health_check(self, endpoint: str) -> 'ServiceProviderInfo':
        """Set the health check endpoint."""
        self.health_check_endpoint = endpoint
        return self
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary for JSON serialization."""
        return {
            "id": self.id,
            "name": self.name,
            "service_type": self.service_type.value,
            "endpoint": self.endpoint,
            "version": self.version,
            "metadata": self.metadata,
            "health_check_endpoint": self.health_check_endpoint,
        }
    
    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> 'ServiceProviderInfo':
        """Create from dictionary."""
        return cls(
            id=data["id"],
            name=data["name"],
            service_type=ServiceType(data["service_type"]),
            endpoint=data["endpoint"],
            version=data["version"],
            metadata=data.get("metadata"),
            health_check_endpoint=data.get("health_check_endpoint"),
        )


class ServiceDiscoveryError(Exception):
    """Base class for service discovery errors."""
    pass


class ConnectionError(ServiceDiscoveryError):
    """Connection to orchestrator failed."""
    pass


class RegistrationError(ServiceDiscoveryError):
    """Service registration failed."""
    pass


class DiscoveryError(ServiceDiscoveryError):
    """Service discovery failed."""
    pass


class Timeout(ServiceDiscoveryError):
    """Operation timed out."""
    def __init__(self, duration: float):
        self.duration = duration
        super().__init__(f"Operation timed out after {duration}s")


@dataclass
class RegisterServiceProviderRequest:
    """Request to register a service provider."""
    request_type: str = "register_service_provider"
    provider_info: Optional[ServiceProviderInfo] = None
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary for JSON serialization."""
        return {
            "request_type": self.request_type,
            "provider_info": self.provider_info.to_dict() if self.provider_info else None,
        }


@dataclass
class RegisterServiceProviderResponse:
    """Response to service provider registration."""
    response_type: str
    success: bool
    error: Optional[str] = None
    provider_id: Optional[str] = None
    
    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> 'RegisterServiceProviderResponse':
        """Create from dictionary."""
        return cls(
            response_type=data["response_type"],
            success=data["success"],
            error=data.get("error"),
            provider_id=data.get("provider_id"),
        )


@dataclass
class DiscoverServiceProvidersRequest:
    """Request to discover service providers."""
    request_type: str = "discover_service_providers"
    service_name: Optional[str] = None
    service_type: Optional[ServiceType] = None
    version: Optional[str] = None
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary for JSON serialization."""
        return {
            "request_type": self.request_type,
            "service_name": self.service_name,
            "service_type": self.service_type.value if self.service_type else None,
            "version": self.version,
        }


@dataclass
class DiscoverServiceProvidersResponse:
    """Response to service provider discovery."""
    response_type: str
    success: bool
    error: Optional[str] = None
    providers: Optional[List[ServiceProviderInfo]] = None
    
    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> 'DiscoverServiceProvidersResponse':
        """Create from dictionary."""
        providers = None
        if data.get("providers"):
            providers = [ServiceProviderInfo.from_dict(p) for p in data["providers"]]
        
        return cls(
            response_type=data["response_type"],
            success=data["success"],
            error=data.get("error"),
            providers=providers,
        )


class ServiceProviderBuilder:
    """Builder for creating service provider information."""
    
    def __init__(self):
        self._id: Optional[str] = None
        self._name: Optional[str] = None
        self._service_type: Optional[ServiceType] = None
        self._endpoint: Optional[str] = None
        self._version: Optional[str] = None
        self._metadata: Dict[str, str] = {}
        self._health_check_endpoint: Optional[str] = None
    
    def with_id(self, id: str) -> 'ServiceProviderBuilder':
        """Set the service provider ID."""
        self._id = id
        return self
    
    def with_name(self, name: str) -> 'ServiceProviderBuilder':
        """Set the service name."""
        self._name = name
        return self
    
    def with_service_type(self, service_type: ServiceType) -> 'ServiceProviderBuilder':
        """Set the service type."""
        self._service_type = service_type
        return self
    
    def with_endpoint(self, endpoint: str) -> 'ServiceProviderBuilder':
        """Set the service endpoint."""
        self._endpoint = endpoint
        return self
    
    def with_version(self, version: str) -> 'ServiceProviderBuilder':
        """Set the service version."""
        self._version = version
        return self
    
    def with_metadata(self, key: str, value: str) -> 'ServiceProviderBuilder':
        """Add metadata."""
        self._metadata[key] = value
        return self
    
    def with_health_check(self, endpoint: str) -> 'ServiceProviderBuilder':
        """Set the health check endpoint."""
        self._health_check_endpoint = endpoint
        return self
    
    def build(self) -> ServiceProviderInfo:
        """Build the service provider info."""
        if not all([self._id, self._name, self._service_type, self._endpoint, self._version]):
            raise ValueError("Missing required fields for service provider")
        
        provider = ServiceProviderInfo(
            id=self._id,
            name=self._name,
            service_type=self._service_type,
            endpoint=self._endpoint,
            version=self._version,
            metadata=self._metadata if self._metadata else None,
            health_check_endpoint=self._health_check_endpoint,
        )
        return provider


class ServiceDiscoveryClient:
    """Client for service discovery operations."""
    
    def __init__(self, channel: TcpChannel):
        self.channel = channel
    
    @classmethod
    async def connect(cls, config: ConnectionConfig) -> 'ServiceDiscoveryClient':
        """Create and connect a new service discovery client."""
        channel = await TcpChannel.connect(config)
        return cls(channel)
    
    async def register_service_provider(self, provider_info: ServiceProviderInfo) -> str:
        """Register a service provider with the orchestrator."""
        try:
            # Send registration request
            request = RegisterServiceProviderRequest(provider_info=provider_info)
            message = Message.new(request.to_dict())
            encoded = message.encode()
            
            await self.channel.send(encoded)
            
            # Wait for response
            response_encoded = await self.channel.receive_with_timeout(10.0)
            response_message = response_encoded.decode()
            response_data = response_message.content
            
            response = RegisterServiceProviderResponse.from_dict(response_data)
            
            if not response.success:
                error_msg = response.error or "Unknown registration error"
                raise RegistrationError(error_msg)
            
            if not response.provider_id:
                raise RegistrationError("Registration succeeded but no provider ID returned")
            
            return response.provider_id
            
        except asyncio.TimeoutError:
            raise Timeout(10.0)
        except MessageError as e:
            raise ConnectionError(f"Channel error: {e}")
    
    async def discover_service_providers(
        self,
        service_name: Optional[str] = None,
        service_type: Optional[ServiceType] = None,
        version: Optional[str] = None
    ) -> List[ServiceProviderInfo]:
        """Discover service providers matching the criteria."""
        try:
            # Send discovery request
            request = DiscoverServiceProvidersRequest(
                service_name=service_name,
                service_type=service_type,
                version=version
            )
            message = Message.new(request.to_dict())
            encoded = message.encode()
            
            await self.channel.send(encoded)
            
            # Wait for response
            response_encoded = await self.channel.receive_with_timeout(10.0)
            response_message = response_encoded.decode()
            response_data = response_message.content
            
            response = DiscoverServiceProvidersResponse.from_dict(response_data)
            
            if not response.success:
                error_msg = response.error or "Unknown discovery error"
                raise DiscoveryError(error_msg)
            
            return response.providers or []
            
        except asyncio.TimeoutError:
            raise Timeout(10.0)
        except MessageError as e:
            raise ConnectionError(f"Channel error: {e}")
    
    async def discover_service_by_name(self, service_name: str) -> List[ServiceProviderInfo]:
        """Discover service providers by name."""
        return await self.discover_service_providers(service_name=service_name)
    
    async def discover_service_by_type(self, service_type: ServiceType) -> List[ServiceProviderInfo]:
        """Discover service providers by type."""
        return await self.discover_service_providers(service_type=service_type)
    
    async def discover_service_by_name_and_type(
        self,
        service_name: str,
        service_type: ServiceType
    ) -> List[ServiceProviderInfo]:
        """Discover service providers by name and type."""
        return await self.discover_service_providers(
            service_name=service_name,
            service_type=service_type
        )
    
    async def close(self) -> None:
        """Close the connection."""
        await self.channel.disconnect()
    
    async def __aenter__(self):
        """Async context manager entry."""
        return self
    
    async def __aexit__(self, exc_type, exc_val, exc_tb):
        """Async context manager exit."""
        await self.close()


# Aliases for backward compatibility
ServiceProvider = ServiceProviderInfo 

File: rust_sdk/src/services/service_discovery.rs
//! Service discovery functionality for PyWatt modules.
//!
//! This module provides functionality for modules to register as service providers
//! and discover other services in the PyWatt ecosystem.

use std::collections::HashMap;
use std::sync::Arc;
use tokio::sync::Mutex;
use tracing::{debug, error, info, warn};
use uuid::Uuid;

use crate::communication::ipc_types::{
    ServiceType, RegisterServiceProviderRequest, RegisterServiceProviderResponse,
    DiscoverServiceProvidersRequest, DiscoverServiceProvidersResponse,
    ServiceProviderInfo, ModuleToOrchestrator, OrchestratorToModule,
};
use crate::message::{Message, MessageMetadata, EncodingFormat};
use crate::communication::{MessageChannel, TcpChannel};
use crate::communication::tcp_types::ConnectionConfig;

/// Result type for service discovery operations.
pub type Result<T> = std::result::Result<T, ServiceDiscoveryError>;

/// Errors that can occur during service discovery operations.
#[derive(Debug, thiserror::Error)]
pub enum ServiceDiscoveryError {
    #[error("Connection error: {0}")]
    ConnectionError(String),
    
    #[error("Serialization error: {0}")]
    SerializationError(String),
    
    #[error("Channel error: {0}")]
    ChannelError(String),
    
    #[error("Timeout error")]
    Timeout,
    
    #[error("Registration failed: {0}")]
    RegistrationFailed(String),
    
    #[error("Discovery failed: {0}")]
    DiscoveryFailed(String),
}

/// Service discovery client for PyWatt modules.
pub struct ServiceDiscoveryClient {
    /// Connection to the orchestrator
    channel: Arc<TcpChannel>,
    /// Registered service providers for this module
    registered_providers: Arc<Mutex<HashMap<String, RegisteredProvider>>>,
}

/// Information about a registered service provider.
#[derive(Debug, Clone)]
pub struct RegisteredProvider {
    pub provider_id: String,
    pub service_type: ServiceType,
    pub name: String,
    pub version: Option<String>,
    pub address: String,
    pub metadata: HashMap<String, String>,
}

impl ServiceDiscoveryClient {
    /// Create a new service discovery client.
    ///
    /// # Arguments
    /// * `config` - Connection configuration for the orchestrator
    ///
    /// # Returns
    /// A new service discovery client
    pub async fn new(config: ConnectionConfig) -> Result<Self> {
        let channel = TcpChannel::connect(config).await
            .map_err(|e| ServiceDiscoveryError::ConnectionError(e.to_string()))?;
        
        Ok(Self {
            channel: Arc::new(channel),
            registered_providers: Arc::new(Mutex::new(HashMap::new())),
        })
    }

    /// Register this module as a service provider.
    ///
    /// # Arguments
    /// * `service_type` - Type of service being provided
    /// * `name` - Human-readable name for the service
    /// * `address` - Network address where the service can be reached
    ///
    /// # Returns
    /// The provider ID if successful
    pub async fn register_service_provider(
        &self,
        service_type: ServiceType,
        name: impl Into<String>,
        address: impl Into<String>,
    ) -> Result<String> {
        let name = name.into();
        let address = address.into();
        
        let request = RegisterServiceProviderRequest {
            service_type: service_type.clone(),
            name: name.clone(),
            version: None,
            address: address.clone(),
            metadata: None,
        };

        let message = ModuleToOrchestrator::RegisterServiceProvider(request);
        let response = self.send_request(message).await?;

        match response {
            OrchestratorToModule::RegisterServiceProviderResponse(resp) => {
                if resp.success {
                    if let Some(provider_id) = resp.provider_id {
                        // Store the registered provider
                        let provider = RegisteredProvider {
                            provider_id: provider_id.clone(),
                            service_type,
                            name,
                            version: None,
                            address,
                            metadata: HashMap::new(),
                        };
                        
                        let mut providers = self.registered_providers.lock().await;
                        providers.insert(provider_id.clone(), provider);
                        
                        info!("Successfully registered as service provider: {}", provider_id);
                        Ok(provider_id)
                    } else {
                        Err(ServiceDiscoveryError::RegistrationFailed("No provider ID returned".to_string()))
                    }
                } else {
                    let error = resp.error.unwrap_or_else(|| "Unknown error".to_string());
                    Err(ServiceDiscoveryError::RegistrationFailed(error))
                }
            }
            _ => Err(ServiceDiscoveryError::RegistrationFailed("Unexpected response type".to_string())),
        }
    }

    /// Register this module as a service provider with additional options.
    ///
    /// # Arguments
    /// * `service_type` - Type of service being provided
    /// * `name` - Human-readable name for the service
    /// * `address` - Network address where the service can be reached
    /// * `version` - Optional version of the service
    /// * `metadata` - Optional metadata about the service
    ///
    /// # Returns
    /// The provider ID if successful
    pub async fn register_service_provider_with_options(
        &self,
        service_type: ServiceType,
        name: impl Into<String>,
        address: impl Into<String>,
        version: Option<String>,
        metadata: Option<HashMap<String, String>>,
    ) -> Result<String> {
        let name = name.into();
        let address = address.into();
        
        let request = RegisterServiceProviderRequest {
            service_type: service_type.clone(),
            name: name.clone(),
            version: version.clone(),
            address: address.clone(),
            metadata: metadata.clone(),
        };

        let message = ModuleToOrchestrator::RegisterServiceProvider(request);
        let response = self.send_request(message).await?;

        match response {
            OrchestratorToModule::RegisterServiceProviderResponse(resp) => {
                if resp.success {
                    if let Some(provider_id) = resp.provider_id {
                        // Store the registered provider
                        let provider = RegisteredProvider {
                            provider_id: provider_id.clone(),
                            service_type,
                            name,
                            version,
                            address,
                            metadata: metadata.unwrap_or_default(),
                        };
                        
                        let mut providers = self.registered_providers.lock().await;
                        providers.insert(provider_id.clone(), provider);
                        
                        info!("Successfully registered as service provider: {}", provider_id);
                        Ok(provider_id)
                    } else {
                        Err(ServiceDiscoveryError::RegistrationFailed("No provider ID returned".to_string()))
                    }
                } else {
                    let error = resp.error.unwrap_or_else(|| "Unknown error".to_string());
                    Err(ServiceDiscoveryError::RegistrationFailed(error))
                }
            }
            _ => Err(ServiceDiscoveryError::RegistrationFailed("Unexpected response type".to_string())),
        }
    }

    /// Discover service providers of a specific type.
    ///
    /// # Arguments
    /// * `service_type` - Type of service to discover
    /// * `all_providers` - Whether to return all providers or just the first healthy one
    ///
    /// # Returns
    /// A list of discovered service providers
    pub async fn discover_service_providers(
        &self,
        service_type: ServiceType,
        all_providers: bool,
    ) -> Result<Vec<ServiceProviderInfo>> {
        let request = DiscoverServiceProvidersRequest {
            service_type,
            all_providers: Some(all_providers),
        };

        let message = ModuleToOrchestrator::DiscoverServiceProviders(request);
        let response = self.send_request(message).await?;

        match response {
            OrchestratorToModule::DiscoverServiceProvidersResponse(resp) => {
                if resp.success {
                    debug!("Discovered {} service providers", resp.providers.len());
                    Ok(resp.providers)
                } else {
                    let error = resp.error.unwrap_or_else(|| "Unknown error".to_string());
                    Err(ServiceDiscoveryError::DiscoveryFailed(error))
                }
            }
            _ => Err(ServiceDiscoveryError::DiscoveryFailed("Unexpected response type".to_string())),
        }
    }

    /// Discover the first healthy service provider of a specific type.
    ///
    /// # Arguments
    /// * `service_type` - Type of service to discover
    ///
    /// # Returns
    /// The first healthy service provider, if any
    pub async fn discover_service_provider(
        &self,
        service_type: ServiceType,
    ) -> Result<Option<ServiceProviderInfo>> {
        let providers = self.discover_service_providers(service_type, false).await?;
        Ok(providers.into_iter().next())
    }

    /// Get all registered service providers for this module.
    ///
    /// # Returns
    /// A list of all registered service providers
    pub async fn get_registered_providers(&self) -> Vec<RegisteredProvider> {
        let providers = self.registered_providers.lock().await;
        providers.values().cloned().collect()
    }

    /// Send a request to the orchestrator and wait for a response.
    async fn send_request(&self, message: ModuleToOrchestrator) -> Result<OrchestratorToModule> {
        // Create message with request ID
        let mut metadata = MessageMetadata::new();
        metadata.id = Some(Uuid::new_v4().to_string());
        
        let msg = Message::with_metadata(message, metadata);
        let encoded = msg.encode()
            .map_err(|e| ServiceDiscoveryError::SerializationError(e.to_string()))?;
        
        // Send the request
        debug!("Sending service discovery request to orchestrator");
        self.channel.send(encoded).await
            .map_err(|e| ServiceDiscoveryError::ChannelError(e.to_string()))?;
        
        // Wait for response with timeout
        debug!("Waiting for service discovery response");
        let response_encoded = tokio::time::timeout(
            std::time::Duration::from_secs(30),
            self.channel.receive()
        ).await
        .map_err(|_| ServiceDiscoveryError::Timeout)?
        .map_err(|e| ServiceDiscoveryError::ChannelError(e.to_string()))?;
        
        // Decode response
        let response_data: OrchestratorToModule = serde_json::from_slice(response_encoded.data())
            .map_err(|e| ServiceDiscoveryError::SerializationError(e.to_string()))?;
        
        Ok(response_data)
    }
}

/// Builder for service provider registration.
pub struct ServiceProviderBuilder {
    service_type: ServiceType,
    name: String,
    address: String,
    version: Option<String>,
    metadata: HashMap<String, String>,
}

impl ServiceProviderBuilder {
    /// Create a new service provider builder.
    ///
    /// # Arguments
    /// * `service_type` - Type of service being provided
    /// * `name` - Human-readable name for the service
    /// * `address` - Network address where the service can be reached
    pub fn new(
        service_type: ServiceType,
        name: impl Into<String>,
        address: impl Into<String>,
    ) -> Self {
        Self {
            service_type,
            name: name.into(),
            address: address.into(),
            version: None,
            metadata: HashMap::new(),
        }
    }

    /// Set the version of the service.
    pub fn with_version(mut self, version: impl Into<String>) -> Self {
        self.version = Some(version.into());
        self
    }

    /// Add metadata to the service.
    pub fn with_metadata(mut self, key: impl Into<String>, value: impl Into<String>) -> Self {
        self.metadata.insert(key.into(), value.into());
        self
    }

    /// Register the service provider using the given client.
    pub async fn register(self, client: &ServiceDiscoveryClient) -> Result<String> {
        client.register_service_provider_with_options(
            self.service_type,
            self.name,
            self.address,
            self.version,
            Some(self.metadata),
        ).await
    }
}

/// Convenience functions for common service types.
impl ServiceDiscoveryClient {
    /// Register as a database service provider.
    pub async fn register_database_provider(
        &self,
        name: impl Into<String>,
        address: impl Into<String>,
    ) -> Result<String> {
        self.register_service_provider(ServiceType::Database, name, address).await
    }

    /// Register as a cache service provider.
    pub async fn register_cache_provider(
        &self,
        name: impl Into<String>,
        address: impl Into<String>,
    ) -> Result<String> {
        self.register_service_provider(ServiceType::Cache, name, address).await
    }

    /// Register as a JWT service provider.
    pub async fn register_jwt_provider(
        &self,
        name: impl Into<String>,
        address: impl Into<String>,
    ) -> Result<String> {
        self.register_service_provider(ServiceType::Jwt, name, address).await
    }

    /// Register as a custom service provider.
    pub async fn register_custom_provider(
        &self,
        service_name: impl Into<String>,
        provider_name: impl Into<String>,
        address: impl Into<String>,
    ) -> Result<String> {
        let service_type = ServiceType::Custom(service_name.into());
        self.register_service_provider(service_type, provider_name, address).await
    }

    /// Discover database service providers.
    pub async fn discover_database_providers(&self, all: bool) -> Result<Vec<ServiceProviderInfo>> {
        self.discover_service_providers(ServiceType::Database, all).await
    }

    /// Discover cache service providers.
    pub async fn discover_cache_providers(&self, all: bool) -> Result<Vec<ServiceProviderInfo>> {
        self.discover_service_providers(ServiceType::Cache, all).await
    }

    /// Discover JWT service providers.
    pub async fn discover_jwt_providers(&self, all: bool) -> Result<Vec<ServiceProviderInfo>> {
        self.discover_service_providers(ServiceType::Jwt, all).await
    }

    /// Discover custom service providers.
    pub async fn discover_custom_providers(
        &self,
        service_name: impl Into<String>,
        all: bool,
    ) -> Result<Vec<ServiceProviderInfo>> {
        let service_type = ServiceType::Custom(service_name.into());
        self.discover_service_providers(service_type, all).await
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::net::SocketAddr;

    #[tokio::test]
    async fn test_service_provider_builder() {
        let builder = ServiceProviderBuilder::new(
            ServiceType::Database,
            "test-db",
            "localhost:5432",
        )
        .with_version("1.0.0")
        .with_metadata("driver", "postgresql")
        .with_metadata("max_connections", "100");

        assert_eq!(builder.service_type, ServiceType::Database);
        assert_eq!(builder.name, "test-db");
        assert_eq!(builder.address, "localhost:5432");
        assert_eq!(builder.version, Some("1.0.0".to_string()));
        assert_eq!(builder.metadata.get("driver"), Some(&"postgresql".to_string()));
        assert_eq!(builder.metadata.get("max_connections"), Some(&"100".to_string()));
    }

    #[test]
    fn test_service_type_display() {
        assert_eq!(ServiceType::Database.to_string(), "database");
        assert_eq!(ServiceType::Cache.to_string(), "cache");
        assert_eq!(ServiceType::Jwt.to_string(), "jwt");
        assert_eq!(ServiceType::Custom("my-service".to_string()).to_string(), "custom:my-service");
    }

    #[test]
    fn test_service_type_custom() {
        let custom_type = ServiceType::custom("my-custom-service");
        assert_eq!(custom_type, ServiceType::Custom("my-custom-service".to_string()));
        assert_eq!(custom_type.to_string(), "custom:my-custom-service");
    }
} 